### PR TITLE
Initial implementation of umv from uutils

### LINKED ISSUE #10446: add uutils/coreutils `mv` into nushell
### Related problem

Following on the momentum of moving the `uu_cp` crate into nushell, we'd like to add the `uu_mv` create too, as `umv` for now. We can use this issue to discuss parameters and other topics.

Looking for volunteers. Reply below and we can assign the issue to you.

### Describe the solution you'd like

Use the `ucp` code as a guide and port `umv` into nushell.

### Describe alternatives you've considered

_No response_

### Additional context and details

_No response_

### PR DESCRIPTION
**Pull request by dmatos2012**
# Description
Hi,
This closes #10446 , wherein we start implementing `mv` from `uutils`. There are some stuff to iron out, particularly
* Decide on behavior from ignored tests 
* Wait for release/PRs to be approved on `uutils` side, but still can be tested for now. See [PR approved](https://github.com/uutils/coreutils/pull/5428), and [pending](https://github.com/uutils/coreutils/pull/5429). 
* `--progress` does not seem to work on `uutils mv` either and have not checked whether certain `X` size has to be achieved in order for it to appear, thus something to investigate as well, but thought it wasnt important enough to not make the PR.

See [issue comment](https://github.com/nushell/nushell/issues/10446#issuecomment-1764497988), on the possible strategy to follow, mainly copy what we did with `ucp`. 

I still left some comments on purpose particularly on tests, which of course would be removed before something is decided here. :)  @fdncred 

# User-Facing Changes

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- [X] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [X] `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- [X] `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- [X] `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library

# After Submitting

### CODE REVIEW

**Comment by fdncred** [comment]
Thanks! Looks like the "PR approved" and "pending" are both landed now. Yay! I guess that means we need to update the hash in the cargo.toml.

Ya, seems like we'd want the progress bar to work if possible.

I'm also wondering if this should use @bobhy's `arg_glob()` that he implemented recently. Not sure `umv` needs it but it would be good to look at it. You can see it in the current `mv.rs`, `ucp.rs`, and `cp.rs`.

It looks like we don't need to have the same discussion about arguments/switches/flags as we did in `ucp` because there really aren't that many.

There seems to be a good amount of tests but it may be good to have some additional tests that test your comments about `paths.is_empty()` and `paths.len() == 1` maybe?

And then just get the ci green. Other than that, this looks real good to me.

**Comment by bobhy** [comment]
Yes, `arg_glob` was intended for all the filesystem commands, I integrated it with uucp already (as well as old mv).

**Comment by dmatos2012** [comment]
yes, I will take a look at the `arg_glob` and implement it here as well, and add tests that you mention @fdncred.

 But can you help me out here since I believe you have a mac. Could you check whether `uutils` also fails on here? [job failed](https://github.com/nushell/nushell/actions/runs/6619794371/job/17980981253?pr=10822), particularly ` commands::move_::umv::mv_change_case_of_file` and ` commands::move_::umv::mv_change_case_of_file`. Since I only have Linux, I could only test here, and all tests pass. I am assuming this comes from `uutils`. 

I

**Comment by fdncred** [comment]
@dmatos2012 Here's some results from my mac.
image
image
image

**Comment by fdncred** [comment]
Let me try again after those recent commits.

**Comment by fdncred** [comment]
They still fail for me on my mac.

**Comment by dmatos2012** [comment]
> They still fail for me on my mac.

Yeah, I did not fix those, could you see if you get the same but on `uutils`? So recreating the testcase with the change_case, I think the originally issue had: 
`mkdir Test`
`mv Test test`

and it would disappear the `Test` dir rather than renamed it.

**Comment by fdncred** [comment]
How's this? This is using my coreutils version installed via brew
```
mv --version  
mv 0.0.22
```
### test_mv_no_clobber

image

### mv_change_case_of_directory

image

### mv_change_case_of_file

image

**Comment by fdncred** [comment]
I'd rather leave this open until the uutils team fixes the issue. Please link the issue so I can see it here when you file it in uutils.

**Comment by fdncred** [comment]
How's this coming @dmatos2012? Wondering if this will make the release on Nov 14th? If so, that means we'd need to land it on or before Nov 12th.

**Comment by dmatos2012** [comment]
> How's this coming @dmatos2012? Wondering if this will make the release on Nov 14th? If so, that means we'd need to land it on or before Nov 12th.

Well, the command is "complete" but couple of things:
* There are failing tests on macOS, and windows but I can only test on linux and here all tests pass.
* Some  decisions that need to be made with the ignored tests now, I could remove them but of course I think the coreteam needs to decide how we want it to behave there. 

Finally, on the biggest problem is that I tried to solve the `changing_case` errors on MacOS, but I think its just simply not possible to make `uutils` tests happy, at least not without signifcant effort. Specifically, because `uutils` has tons of tests, and slight change of behavior makes the tests break and then keeping compatibility accross OS can be a nightmare :(

Particularly, is that windows and macOS are case insensitive, so from `uutils` there is no way to distinguish between 
Case A:
```
touch samefile
mv samefile Samefile
```
Desired behavior: Rename samefile to Samefile
Actual behavior: Cannot rename since samefile *is* same file as Samefile

Case B:
```
touch samefile
ln samefile Samefile
mv samefile Samefile
```
Desired behavior: Error out since there are the same files
Actual behavior: Cannot rename since samefile is same file as Samefile

The problem is from `uutils` point of view, is that there is no obvious  way to distinguish case A) from case B) as in both cases:
* Target exists
* They are the same file
* !src.eq(target)

So either we error out with `mv: cannot move: samefile and Samefile are the same file` or we try to rename by first removing samefile, and  then creating Samefile. See my [PR in uutils](https://github.com/uutils/coreutils/pull/5495). for more details.

Conclusion: Either we handle the change case directory in nushell itself before `uutils` or allow it on `linux` and error out on both `macOS and windows`. I do think it shouldnt be a big deal as if someone really wants to rename a directory in case-insensitive file systems they could do `mv samefile samefile2 && mv samefile2 Samefile`, just small extra step.

Once these are resolved we can continue I think. There are other two `macOS` tests failing, but I really struggle to make them pass myself as I do not see whats wrong with them, so some help here would be greatly appreciated to get this through.

**Comment by fdncred** [comment]
Thanks for the update! I'll see if anyone else has bandwidth to help. I'll be out of pocket for a few weeks probably (surgery).

**Comment by dmatos2012** [comment]
> Thanks for the update! I'll see if anyone else has bandwidth to help. I'll be out of pocket for a few weeks probably (surgery).

Yes, I think its most with knowing what behaviors we want with nushell vs whats currently with `uutils` and someone with a mac. But more importantly, wish you the best in your surgery. Speedy recovery and hope to see you here soon helping with my silly questions! :)

**Comment by fdncred** [comment]
Hopefully we can get this in version 0.88. We have 4 weeks until the release. Would love to get uu_mv into nushell.

**Comment by tertsdiepraam** [comment]
So, we need to be able to rename `a` to `A` on mac? Can we confirm that Mac's `mv` does this as well?

**Comment by fdncred** [comment]
@tertsdiepraam I can confirm that MacOS's `mv` does this.
image
I can try GNU on MacOS if you can point me to where I can download it.

**Comment by tertsdiepraam** [comment]
Interesting! A fairly simple way to install the GNU version is to use homebrew: https://formulae.brew.sh/formula/coreutils. It should then be installed as `gmv`.

**Comment by fdncred** [comment]
I think I did it right. I had to unlink uutils-coreutils but after I did that I installed coreutils and tried again.
image

**Comment by tertsdiepraam** [comment]
Right, so GNU doesn't do it? That makes it a bit complicated. We could also see if there's been any talk on the GNU side of things here. I'll investigate a bit.

**Comment by fdncred** [comment]
> Right, so GNU doesn't do it? That makes it a bit complicated. We could also see if there's been any talk on the GNU side of things here. I'll investigate a bit.

@tertsdiepraam any update on this? I'd love to get uu_mv landed.

**Comment by tertsdiepraam** [comment]
I forgot about this 😅 

But I just checked and found these discussions:

- https://lists.gnu.org/archive/html/bug-coreutils/2007-08/msg00039.html
- https://lists.gnu.org/archive/html/bug-coreutils/2007-08/msg00042.html
- https://lists.gnu.org/archive/html/bug-coreutils/2007-06/msg00207.html
- https://lists.gnu.org/archive/html/bug-coreutils/2007-03/msg00218.html

I'm not really getting a clear picture of whether they want to do this. I think the main problem with implementing is is that the we can't just assume the sensitivity based on the OS, but need to look at the file system instead. I do agree that this behavior would be nice to have though.

**Comment by fdncred** [comment]
I'd also be fine with landing something close and continuing to work on this case-sensitivity approach.

**Comment by fdncred** [comment]
any updates on this front, we just had another release without our friendly neighborhood uu_mv 😭

**Comment by tertsdiepraam** [comment]
I'm not really sure how to continue to be honest. I don't really have the bandwidth to work on it at the moment either. But if @dmatos2012 or anyone else has an idea for solving this I'd be happy to review!

**Comment by fdncred** [comment]
Thanks @tertsdiepraam! as always appreciate your feedback. I'm up for whatever can get this landed in a step towards the right direction. It doesn't have to be perfect. @dmatos2012 do you have any bandwidth coming up that you can work on this a bit more and get to something usable?

**Comment by dmatos2012** [comment]
Hi @fdncred. To be honest , I think im on the same boat as @tertsdiepraam without knowing how to proceed further with this. Although I probably have more bandwidth for this , I do not think I can contribute further on this issue (case sensitive problems that stops the PR). 

I would gladly help with porting another `uutils` in the future, or even other `umv` issue,  but not sure how to proceed with this one :(

**Comment by fdncred** [comment]
@dmatos2012 I guess what I'm saying is for nushell to accept the GNU mv and uu_mv functionality as it is and document it. So, instead of supporting `mv a A` like the built-in macos does, just give return an error like it does now. Maybe just say `mv: cannot move: samefile and Samefile are the same file` and then add `Your OS's `mv command may allow moving this way. If you want to rename files with the same file name but different case use ^mv.` or some message that makes sense.

**Comment by dmatos2012** [comment]
Ok,  I will rebase this, and see the current failing tests and fix them according to the current `uutils mv` behavior. Bit of holiday season, but will give it a try soon :) , hope thats ok @fdncred

**Comment by fdncred** [comment]
sounds great @dmatos2012. as always, we appreciate your help.

**Comment by kubouch** [comment]
Thanks for the mv! Please also ensure that absolute paths are sent to uutils to avoid `mv`-after-`cd` errors coming from expanding paths from `std::env` like in https://github.com/nushell/nushell/issues/10832. See also https://github.com/nushell/nushell/pull/11080 and https://github.com/nushell/nushell/pull/11152. Ideally, a test covering this case would be great. This applies to all uutils commands that do something with files, essentially.

File: crates/nu-command/src/filesystem/umv.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                            msg: format!("error {} in path {}", e.error(), e.path().display()),
+                            span: p.span,
+                        });
+                    }
+                }
+            }
+            files.append(&mut app_vals);
+        }
+        // Add back the target after globbing
+        let target = paths.last().expect("Should not be reached");
```
Comment (kubouch) [inline crates/nu-command/src/filesystem/umv.rs:147]: Could you refactor the `expect()` into some error? We try to minimize potential panics as much as possible in Nushell.

**Comment by kubouch** [comment]
OK, error is better. Sorry for the nitpicking, but could you put there an error that takes Span? There might be `NushellFailedSpanned` or similar, you can use `call.head` as the span. Errors without spans are strongly discouraged in Nushell as they make code very hard to debug.

**Comment by dmatos2012** [comment]
@kubouch thanks for the feedback. I still gotta iron out some of the failing tests on other platforms just to make the CI pass as we know some of the behavior is changing. However, I keep getting the same test failing, when locally on my ubuntu passes just fine. Are tests being cached on CI rerun?

**Comment by kubouch** [comment]
I don't think they're cached in the CI. Each CI run should be fresh.

**Comment by fdncred** [comment]
@dmatos2012 we appreciate your continued commits here trying to resolve the ci issues. i wish there was an easier way.

**Comment by dmatos2012** [comment]
> @dmatos2012 we appreciate your continued commits here trying to resolve the ci issues. i wish there was an easier way.

Could you run those two failing tests on your Mac if possible? The errors on both are correct but somehow I think some new lines are messing up the `err.contains(actual.err)` part


So maybe you have a clever way to fix it? I think just because the paths are way too long. I take suggestions :)

**Comment by fdncred** [comment]
@dmatos2012 This is what I get when I run it on my mac
```
failures:

---- tests::test_config_path::test_alternate_config_path stdout ----
thread 'tests::test_config_path::test_alternate_config_path' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::test_config_path::test_default_config_path stdout ----
thread 'tests::test_config_path::test_default_config_path' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)

---- tests::test_custom_commands::deprecated_boolean_flag stdout ----
thread 'tests::test_custom_commands::deprecated_boolean_flag' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)

---- tests::test_custom_commands::infinite_recursion_does_not_panic stdout ----
thread 'tests::test_custom_commands::infinite_recursion_does_not_panic' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)

---- tests::test_custom_commands::override_table_eval_file stdout ----
thread 'tests::test_custom_commands::override_table_eval_file' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)

---- tests::test_spread::deprecate_implicit_spread_for_externals stdout ----
thread 'tests::test_spread::deprecate_implicit_spread_for_externals' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)

---- tests::test_spread::spread_external_args stdout ----
thread 'tests::test_spread::spread_external_args' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)

---- tests::test_spread::spread_non_list_args stdout ----
stdout:
stderr: Error: nu::shell::cannot_spread_as_list

  × Not a list
   ╭─[/private/var/folders/m1/x0gp9jy51s146ttxtz7jlpxh0000gn/T/.tmpmv8Vsk:1:1]
 1 │ echo ...(1)
   ·         ─┬─
   ·          ╰── cannot spread value
   ╰────
  help: Only lists can be spread inside lists and command calls. Try
        converting the value to a list before spreading.


Expected error to contain: cannot spread value
thread 'tests::test_spread::spread_non_list_args' panicked at /Users/fdncred/src/nushell/crates/nu-test-support/src/macros.rs:270:21:
Can't run test "/Users/fdncred/src/nushell/target/debug/nu" No such file or directory (os error 2)


failures:
    tests::test_config_path::test_alternate_config_path
    tests::test_config_path::test_default_config_path
    tests::test_custom_commands::deprecated_boolean_flag
    tests::test_custom_commands::infinite_recursion_does_not_panic
    tests::test_custom_commands::override_table_eval_file
    tests::test_spread::deprecate_implicit_spread_for_externals
    tests::test_spread::spread_external_args
    tests::test_spread::spread_non_list_args

test result: FAILED. 562 passed; 8 failed; 4 ignored; 0 measured; 0 filtered out; finished in 5.92s

error: test failed, to rerun pass `-p nu --bin nu`
```

**Comment by fdncred** [comment]
My last post is another problem. Here's the current test run.
```
failures:

---- commands::move_::umv::errors_if_moving_to_itself stdout ----
=== stderr
Error:   × cannot move '/private/var/folders/m1/x0gp9jy51s146ttxtz7jlpxh0000gn/
  │ T/.tmprxTqjV/umv_test_10_4/mydir' to a subdirectory of itself, '/private/
  │ var/folders/m1/x0gp9jy51s146ttxtz7jlpxh0000gn/T/.tmprxTqjV/umv_test_10_4/
  │ mydir/mydir_2///private/var/folders/m1/x0gp9jy51s146ttxtz7jlpxh0000gn/
  │ T/.tmprxTqjV/umv_test_10_4/mydir'



thread 'commands::move_::umv::errors_if_moving_to_itself' panicked at crates/nu-command/tests/commands/move_/umv.rs:301:13:
Failure: expected err was
cannot move '/private/var/folders/m1/x0gp9jy51s146ttxtz7jlpxh0000gn/T/.tmprxTqjV/umv_test_10_4/mydir'
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- commands::move_::umv::test_mv_no_clobber stdout ----
=== stderr
Error:   × not replacing '/private/var/folders/m1/x0gp9jy51s146ttxtz7jlpxh0000gn/
  │ T/.tmp8EyJBM/umv_test_13/test_mv_no_clobber_file_b'



thread 'commands::move_::umv::test_mv_no_clobber' panicked at crates/nu-command/tests/commands/move_/umv.rs:564:13:
Failure: stderr was
not replacing '/private/var/folders/m1/x0gp9jy51s146ttxtz7jlpxh0000gn/T/.tmp8EyJBM/umv_test_13/test_mv_no_clobber_file_b'


failures:
    commands::move_::umv::errors_if_moving_to_itself
    commands::move_::umv::test_mv_no_clobber

test result: FAILED. 1403 passed; 2 failed; 27 ignored; 0 measured; 0 filtered out; finished in 22.21s

error: test failed, to rerun pass `-p nu-command --test main`
```

**Comment by fdncred** [comment]
@dmatos2012 I changed the tests and got it to pass on macos.
```rust
#[test]
fn errors_if_moving_to_itself() {
    Playground::setup("umv_test_10_4", |dirs, sandbox| {
        sandbox.mkdir("mydir").mkdir("mydir/mydir_2");

        let actual = nu!(
            cwd: dirs.test(),
            "umv mydir mydir/mydir_2/"
        );
        let expected_error = "cannot move";
        if !actual.err.contains(expected_error) {
            panic!("Failure: expected err was \n{}", expected_error);
        }
        assert!(actual.err.contains("to a subdirectory of itself"));
    });
}
```
and
```rust
#[test]
fn test_mv_no_clobber() {
    Playground::setup("umv_test_13", |dirs, sandbox| {
        let file_a = "test_mv_no_clobber_file_a";
        let file_b = "test_mv_no_clobber_file_b";
        sandbox.with_files(vec![EmptyFile(file_a)]);
        sandbox.with_files(vec![EmptyFile(file_b)]);

        let actual = nu!(
            cwd: dirs.test(),
            "umv -n {} {}",
            file_a,
            file_b,
        );

        let expected_error = "not replacing";
        if !actual.err.contains(expected_error) {
            panic!("Failure: stderr was \n{}", expected_error);
        }
    })
}
```

**Comment by dmatos2012** [comment]
Thanks to @fdncred and his Mac,  finally the tests are all passing now, and I have cleaned it up.  I think this is now ready @fdncred . I already have made the changes @kubouch suggested the previous time.

**Comment by fdncred** [comment]
Thanks @dmatos2012! Glad to have this landed as `umv` for further testing.

**Comment by WindSoilder** [comment]
It seems that `umv` doesn't fix #11127 either :-(  It's failed with no output
